src/hotspot/share/utilities/packedTable.cpp

```diff
@@ -0,0 +1,120 @@
+ */
+#include <cstring>
+
+// MSVC++ does not have an appropriate builtin, use library version
+#ifdef _MSC_VER
+# define __builtin_memcpy memcpy
+#endif
+
+#include "utilities/align.hpp"
+#include "utilities/count_leading_zeros.hpp"
+#include "utilities/packedTable.hpp"
+
+// The thresholds are inclusive, and in practice the limits are rounded
+// to the nearest power-of-two - 1.
+// Based on the max_key and max_value we figure out the number of bits required to store
+// key and value; imagine that only as bits (not aligned to byte boundary... yet).
+// Then we concatenate the bits for key and value, and 'add' 1-7 padding zeroes
+// (high-order bits) to align on bytes.
+// In the end we have each element in the table consuming 1-8 bytes (case with 0 bits for key
+// is ruled out).
+PackedTableBase::PackedTableBase(uint32_t max_key, uint32_t max_value) {
+  unsigned int key_bits = max_key == 0 ? 0 : 32 - count_leading_zeros(max_key);
+  unsigned int value_bits = max_value == 0 ? 0 : 32 - count_leading_zeros(max_value);
+  _element_bytes = align_up(key_bits + value_bits, 8) / 8;
+  // shifting left by 32 is undefined behaviour, and in practice returns 1
+  _key_mask = key_bits >= 32 ? -1 : (1U << key_bits) - 1;
+  _value_shift = key_bits;
+  _value_mask = value_bits >= 32 ? -1 : (1U << value_bits) - 1;
+  guarantee(_element_bytes > 0, "wouldn't work");
+  assert(_element_bytes <= sizeof(uint64_t), "shouldn't happen");
+}
+
+// Note: we require the supplier to provide the elements in the final order as we can't easily sort
+// within this method - qsort() accepts only pure function as comparator.
+void PackedTableBuilder::fill(u1* table, size_t table_length, Supplier &supplier) const {
+  uint32_t key, value;
+  size_t offset = 0;
+  for (; offset + sizeof(uint64_t) <= table_length && supplier.next(&key, &value); offset += _element_bytes) {
+    assert((key & ~_key_mask) == 0, "key out of bounds");
+    assert((value & ~_value_mask) == 0, "value out of bounds: %x vs. %x (%x)", value, _value_mask, ~_value_mask);
+    uint64_t element = static_cast<uint64_t>(key) | (static_cast<uint64_t>(value) << _value_shift);
+    __builtin_memcpy(table + offset, &element, _element_bytes);
+  }
+  // last bytes
+  for (; offset < table_length && supplier.next(&key, &value); offset += _element_bytes) {
+    uint64_t element = static_cast<uint64_t>(key) | (static_cast<uint64_t>(value) << _value_shift);
+    for (unsigned int i = 0; i < _element_bytes; ++i) {
+      table[offset + i] = static_cast<u1>(0xFF & (element >> (8 * i)));
+    }
+  }
+
+  assert(offset == table_length, "Did not fill whole array");
+  assert(!supplier.next(&key, &value), "Supplier has more elements");
+}
+
+uint64_t PackedTableLookup::read_element(size_t offset) const {
+  uint64_t element = 0;
+  __builtin_memcpy(&element, _table + offset, _element_bytes);
+  assert((element & ~((uint64_t) _key_mask | ((uint64_t) _value_mask << _value_shift))) == 0, "read too much");
+  return element;
+}
```

Comment:
Since `_element_bytes` can be smaller than 8, memcpy will not work on big endian.

Why are you trying to optimize this? This PR already cuts down the iterations from O(n) to O(log(n)). You are already doing a lot at each iteration -- decoding the name_index and signature_index from the unsigned5 stream, looking up the symbols from the constant pool, etc.  So a few bit operations in read_element isn't going to make any substantial difference:

```
uint64_t element = 0;
for (int i = 0; i < _elements_bytes; i++) {
   element <<= 8;
   element |= _table[offset++]; // Need to rewrite fill() accordingly.
}
```